Generate kubectl documentation with kubectl not 3 other binaries

eparis opened the pull request:
This adds 3 new subcommands to kubectl. `kubectl genmd` `kubectl genman` kubectl genbash` which will output the documentation for kubectl. cobra is automatically generating these things, so we can also stop doing those things ourselves (for man pages)

eparis committed:
Code to autogenerate markdown, man pages, and bash completions

eparis committed:
Do kubectl generation in kubectl instead of in 3 seperate binaries

eparis commented:
@mikedanese this is part #2 of breaking up #13018 

It is based on top of #13613

deads2k commented on hack/lib/util.sh:
```diff
@@ -48,6 +48,36 @@ kube::util::wait_for_url() {
+
+# Opposite of kube::util::ensure-temp-dir()
+kube::util::cleanup-temp-dir() {
+  rm -rf "${KUBE_TEMP}" || true
```
I'm not up to date with all the kube scripts, but this fails on an unbound `KUBE_TEMP`, correct?  Protect inside of an `if`?

eparis replied:
It would if you called `kube::util::clenup-temp-dir()` without having called `kube::util::ensure-temp-dir()`  Doing so feels like a coding problem that should cause it to fail. In fact, the only caller is the trap from `ensure-temp-dir` so I think we are safe, no?

deads2k replied:
> It would if you called kube::util::clenup-temp-dir() without having called kube::util::ensure-temp-dir() Doing so feels like a coding problem that should cause it to fail. In fact, the only caller is the trap from ensure-temp-dir so I think we are safe, no?

You've gone to the effort of `|| true` to prevent failures, I just expected this to go one way or the other ("you must have preconditions correct" or "cleanup will never fail").  I don't feel strongly, I just got bit by a similar bash function in openshift.

eparis replied:
hmmm, my worry was if someone created a file we couldn't delete, that shouldn't happen. I'll kill the || true and go with it...

deads2k commented on hack/lib/util.sh:
```diff
@@ -124,60 +155,51 @@ kube::util::wait-for-jobs() {
-  # generate the new files
-  ${cmd} "${tmpdir}"
+# $1 is the directory to put the generated documents in
+kube::util::gen-docs() {
```
New name to reflect what this does, since it's no longer generic.

eparis replied:
Hmmmm, suggestions? It does now gen-docs...

deads2k replied:
> Hmmmm, suggestions? It does now gen-docs...

`gen-kubectl-docs`?  It is now specific to that binary, correct?

eparis replied:
End goal is like https://github.com/kubernetes/kubernetes/pull/13018/files#diff-ef2873ebed657da627ed30f61dea2964R158 where it does it for everything...

deads2k commented on hack/lib/util.sh:
```diff
@@ -124,60 +155,51 @@ kube::util::wait-for-jobs() {
-  mkdir -p "${tmpdir}"
-  # generate the new files
-  ${cmd} "${tmpdir}"
+# $1 is the directory to put the generated documents in
```
Update doc to indicate what you're generating for.

deads2k commented on cmd/genutils/genutils.go:
```diff
@@ -39,3 +41,65 @@ func OutDir(path string) (string, error) {
+		path = args[0]
+	} else if len(args) > 1 {
+		fmt.Fprintf(os.Stderr, "usage: COMMAND [output directory]\n")
+		os.Exit(1)
```
Harsh error case for such an innocuous method name.  Return an error and the keep the exits/fatals in the top level of the command.

deads2k commented on docs/man/man1/kubectl-annotate.1:
```diff
@@ -30,7 +30,7 @@ resourcequotas (quota) or secrets.
 .SH OPTIONS
 .PP
-\fB\-\-all\fP=false
+\fB\-\-all\fP[=false]
```
What do these brackets do?

eparis replied:
`[]` in man pages indicate that the argument is optional.

deads2k replied:
> [] in man pages indicate that the argument is optional.

ok

deads2k commented on docs/man/man1/kubectl-annotate.1:
```diff
@@ -1,4 +1,4 @@
-.TH "KUBERNETES" "1" " kubernetes User Manuals" "Eric Paris" "Jan 2015"  ""
+.TH "KUBERNETES" "1" ""  ""
```
Is the " kubernetes User Manuals" important?

eparis replied:
The date shows up at the end, I'll look into making both of those things possible.

deads2k commented on docs/man/man1/kubectl-config-view.1:
```diff
@@ -174,9 +174,9 @@ $ kubectl config view \-o template \-\-template='{{range .users}}{{ if eq .name
 
 .SH SEE ALSO
 .PP
-\fBkubectl\-config(1)\fP,
```
This change is interesting.  Without the dash, the link is broken, right?

eparis replied:
Definitely broken...

deads2k commented on docs/man/man1/kubectl-expose.1:
```diff
@@ -27,11 +27,7 @@ re\-use the labels from the resource it exposes.
     Synonym for \-\-target\-port
 
 .PP
-\fB\-\-create\-external\-load\-balancer\fP=false
```
Is this option gone now and somehow our verify script didn't catch it?

eparis replied:
It's deprecated. But we were doing the man page generation wrong. Now we are using cobra which does it right.

deads2k commented on docs/man/man1/.files_generated:
```diff
@@ -1,31 +0,0 @@
-kubectl-annotate.1
```
Was this file used for anything?  It doesn't seem terribly useful, but I don't know it would have existed initially.

eparis replied:
It is/was used the same as .generated_docs after this PR. It was a way to know which files in the existing repo need to be removed before new docs are put in. If you removed the kubectl-annotate subcommand (from kubectl) and just put new docs in place it would leave the old kubectl-annotate files in the repo.

deads2k commented on hack/after-build/update-generated-docs.sh:
```diff
@@ -23,17 +23,23 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
+  rm "${KUBE_ROOT}/${file}" 2>/dev/null || true
+done <"${KUBE_ROOT}/.generated_docs"
+
+# the shopt is so that we get .files_generated from the glob.
```
Check this comment.  We don't seem to have those files.

eparis replied:
right, now it is .generated_docs (or similar)

deads2k commented on hack/after-build/verify-generated-docs.sh:
```diff
@@ -59,12 +51,15 @@ if [[ $ret -gt 1 ]]; then
-echo "diffing ${DOCROOT} against freshly generated docs"
-diff -Naupr "${DOCROOT}" "${TMP_DOCROOT}" && ret=0 || ret=$?
-rm -rf "${_tmp}"
+kube::util::ensure-temp-dir
```
marker for where I left off.

deads2k commented on cmd/genutils/genutils.go:
```diff
@@ -39,3 +41,65 @@ func OutDir(path string) (string, error) {
+	return outDir
+}
+
+func AddGeneratedDocsCommands(cmd *cobra.Command) {
```
godoc here explaining that these are going to be attached to the a real command, but the `Deprecated` string will prevent them from showing up in help.  

They are prevented from showing up in help, right?

eparis replied:
correct, added todo to convert to https://github.com/spf13/cobra/pull/143 after it lands.

eparis committed:
Update github.com/spf13/cobra and spf13/pflag

bump(github.com/spf13/pflag) 53d96508f5c9311f5dade714e1de7bd6a9d2af4a
bump(github.com/spf13/cobra) 5a13a75b13d132dc7d07dd19fd260257cd9c0290

This adds support for both man page generation and for hidden commands

eparis committed:
Update docs after changes

eparis commented:
Hopefully all comments (thus far) addressed.

deads2k commented:
Fix your bump commit to be "bump(github.com/spf13/cobra): <commit>" so its easier to find.

deads2k commented:
You've made sure that the verify still functions properly?  If so,

lgtm once you've updated the bump commit name, squashed, and gone green.

deads2k commented:
@eparis bump.

Still not green.  Has anything substantive changed since I reviewed?

eparis commented:
rebased, no major changes. addressed all of (I hope) @deads2k comments. In accordance with with LGTM above, I'm adding the label.

alex-mohr commented:
@eparis shippable/travis unhappy:
hack/verify-godeps.sh master
Starting to download all kubernetes godeps. This takes a while
package github.com/prometheus/client_golang/extraction
	imports github.com/prometheus/client_golang/extraction
	imports github.com/prometheus/client_golang/extraction: cannot find package "github.com/prometheus/client_golang/extraction" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/github.com/prometheus/client_golang/extraction (from $GOROOT)
	/tmp/gopath.FRVhEt/src/github.com/prometheus/client_golang/extraction (from $GOPATH)
godep: restore: exit status 1
package github.com/prometheus/client_golang/model
	imports github.com/prometheus/client_golang/model
	imports github.com/prometheus/client_golang/model: cannot find package "github.com/prometheus/client_golang/model" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/github.com/prometheus/client_golang/model (from $GOROOT)
	/tmp/gopath.FRVhEt/src/github.com/prometheus/client_golang/model (from $GOPATH)
godep: restore: exit status 1
!!! Error in hack/verify-godeps.sh:46
  '"${GODEP}" restore' exited with status 1
Call stack:
  1: hack/verify-godeps.sh:46 main(...)

eparis commented:
That wasn't me  :-(   must mean promethius changed and it is no longer possible for a new user to do `godep restore`.  I'll take a look.  Uggggh.

eparis commented:
caused by: https://github.com/prometheus/client_golang/pull/173

every single update to godeps is going to fail from now on...

thockin commented:
@eparis what is teh state of the prometheus/godep mess?  Is this going to be fixed today, or should we manually merge?

eparis commented:
https://github.com/kubernetes/kubernetes/pull/14179

deads2k commented:
@eparis still no substantive changes?  I know that command changes have gone in, so be sure you regenerated docs and all.

deads2k commented:
@eparis is this dead or alive?

eparis commented:
assigned to me until I rebase.

bgrant0607 commented:
cc @kubernetes/kubectl

fejta commented:
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!
